conn.go

```diff
@@ -311,25 +309,36 @@ func (c *Conn) initTLSConfig() {
 
 // start establishes the connection and begins multiplexing network IO.
 // It is an error to call Start() on a connection that's been closed.
-func (c *Conn) start(deadline time.Time) error {
-	// set connection establishment deadline
-	_ = c.net.SetDeadline(deadline)
-
-	// run connection establishment state machine
-	for state := c.negotiateProto; state != nil; {
-		var err error
-		state, err = state()
-		// check if err occurred
-		if err != nil {
-			close(c.txDone) // close here since connWriter hasn't been started yet
-			close(c.rxDone)
-			_ = c.Close()
-			return err
-		}
+func (c *Conn) start(ctx context.Context) (err error) {
+	// if the context has a deadline or is cancellable, start the interruptor goroutine.
+	// this will close the underlying net.Conn in response to the context.
+
+	if ctx.Done() != nil {
```

Comment:
This seems pretty close to just having a wrapper context. It'd simplify it a little, I think.

Reply:
Not sure I follow, can you please clarify?

Reply:
Something like this?

```go
subCtx, cancelSub := context.WithCancel(ctx)
defer cancelSub()

defer func() {
    <-subCtx    

    if ctx.Err() != nil {
        c.closeDuringStart()
        err = ctx.Err()
    }
}()
```

Reply:
I see.  In this case, we don't want to enable the interrupter goroutine if the provided context is not cancellable/has no deadline.  E.g. if a caller passes `context.Background()` for `ctx` then `ctx.Done()` returns `nil`.

Reply:
Yeah, then it just doesn't do anything (ie, waste). No biggie, just thought it was an interesting pattern.